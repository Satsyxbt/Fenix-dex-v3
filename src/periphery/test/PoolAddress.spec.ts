import { ZeroAddress } from 'ethers';
import { ethers } from 'hardhat';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';

import { PoolAddressTest } from '../typechain';
import { POOL_BYTECODE_HASH } from './shared/computePoolAddress';
import { expect } from './shared/expect';
import snapshotGasCost from './shared/snapshotGasCost';

describe('PoolAddress', () => {
  let poolAddress: PoolAddressTest;

  const poolAddressTestFixture = async () => {
    const poolAddressTestFactory = await ethers.getContractFactory('PoolAddressTest');
    return (await poolAddressTestFactory.deploy()) as any as PoolAddressTest;
  };

  beforeEach('deploy PoolAddressTest', async () => {
    poolAddress = await loadFixture(poolAddressTestFixture);
  });

  describe('#POOL_INIT_CODE_HASH', () => {
    it('equals the hash of the pool bytecode', async () => {
      expect(await poolAddress.POOL_INIT_CODE_HASH()).to.eq(POOL_BYTECODE_HASH);
    });
  });

  describe('#computeAddress', () => {
    it('all arguments equal zero', async () => {
      await expect(poolAddress.computeAddress(ZeroAddress, ZeroAddress, ZeroAddress, 0)).to.be.reverted;
    });

    it('matches example from core repo', async () => {
      expect(
        await poolAddress.computeAddress(
          '0x5FbDB2315678afecb367f032d93F642f64180aa3',
          '0x1000000000000000000000000000000000000000',
          '0x2000000000000000000000000000000000000000',
          250
        )
      ).to.matchSnapshot();
    });

    it('token argument order cannot be in reverse', async () => {
      await expect(
        poolAddress.computeAddress(
          '0x5FbDB2315678afecb367f032d93F642f64180aa3',
          '0x2000000000000000000000000000000000000000',
          '0x1000000000000000000000000000000000000000',
          3000
        )
      ).to.be.reverted;
    });

    it('gas cost [ @skip-on-coverage ]', async () => {
      await snapshotGasCost(
        poolAddress.getGasCostOfComputeAddress(
          '0x5FbDB2315678afecb367f032d93F642f64180aa3',
          '0x1000000000000000000000000000000000000000',
          '0x2000000000000000000000000000000000000000',
          3000
        )
      );
    });
  });
});
