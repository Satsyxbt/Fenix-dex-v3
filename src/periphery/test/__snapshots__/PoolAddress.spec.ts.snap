// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolAddress #computeAddress gas cost [ @skip-on-coverage ] 1`] = `625`;

exports[`PoolAddress #computeAddress matches example from core repo 1`] = `"0x84d429Ad2ffA3967a7607e637973f4Dc97871166"`;
